src/Simplify.elm

```diff
@@ -1154,100 +1198,158 @@ declarationVisitor declarationNode context =
                             context.inferredConstants
                     in
-                    { context | inferredConstants = ( inferredConstants, previous :: previousStack ) }
-
-                Nothing ->
-                    context
+                    { context
+                        | inferredConstants = ( inferredConstants, previous :: previousStack )
+                    }
     in
-    if List.member (Node.range node) newContext.rangesToIgnore then
-        ( [], newContext )
+    if RangeDict.member (Node.range node) context.rangesToIgnore then
+        ( [], contextWithInferredConstants )
 
     else
         let
+            contextWithInferredConstantsAndLocalBindings : ModuleContext
+            contextWithInferredConstantsAndLocalBindings =
+                { contextWithInferredConstants
+                    | localBindings =
+                        RangeDict.union context.localBindings (expressionSurfaceBindings node)
+                }
+
             { errors, rangesToIgnore, rightSidesOfPlusPlus, inferredConstants } =
-                expressionVisitorHelp node newContext context.importLookup
+                expressionVisitorHelp node contextWithInferredConstantsAndLocalBindings
         in
         ( errors
-        , { newContext
-            | rangesToIgnore = rangesToIgnore ++ newContext.rangesToIgnore
-            , rightSidesOfPlusPlus = rightSidesOfPlusPlus ++ newContext.rightSidesOfPlusPlus
-            , inferredConstantsDict = List.foldl (\( range, constants ) acc -> RangeDict.insert range constants acc) newContext.inferredConstantsDict inferredConstants
+        , { contextWithInferredConstantsAndLocalBindings
+            | rangesToIgnore = RangeDict.union context.rangesToIgnore rangesToIgnore
+            , rightSidesOfPlusPlus = RangeDict.union rightSidesOfPlusPlus context.rightSidesOfPlusPlus
+            , inferredConstantsDict =
+                RangeDict.union context.inferredConstantsDict (RangeDict.fromList inferredConstants)
           }
         )
 
 
+expressionSurfaceBindings : Node Expression -> RangeDict (Set String)
+expressionSurfaceBindings expression =
+    RangeDict.map (\collectBindings -> collectBindings ())
+        (expressionSurfaceBindingIntroductions expression)
+
+
+expressionSurfaceRangesAfterPatterns : Node Expression -> RangeDict ()
+expressionSurfaceRangesAfterPatterns expression =
+    RangeDict.map (\_ -> ())
+        (expressionSurfaceBindingIntroductions expression)
+
+
+{-| Whenever you add ranges on expression enter, the same ranges should be removed on expression exit.
+Having one function finding unique ranges and a function for extracting bindings there ensures said consistency.
+
+An alternative approach would be to use some kind of tree structure
+with parent and sub ranges and bindings as leaves (maybe a "trie", tho I've not seen one as an elm package).
+
+Removing all bindings for an expression's range on leave would then be trivial
```

Comment:
What I usually do is to have a stack of scopes, and a separate RangeDict to know when to pop the stack in the exit visitor.

Another possibility would be to have to store the elements collected by `expressionSurfaceBindings` and store them in a RangeDict. Whenever the expression corresponding to the larger scope is encountered, we can remove all what was contained.